packages/teamscale-javascript-instrumenter/src/instrumenter/Postprocessor.ts

```diff
@@ -0,0 +1,364 @@
+import { parse } from '@babel/parser';
+import generate from '@babel/generator';
+import traverse, { NodePath } from '@babel/traverse';
+import {
+	Node,
+	CallExpression,
+	Identifier,
+	isCallExpression,
+	isIdentifier,
+	isMemberExpression,
+	NumericLiteral,
+	SourceLocation,
+	UpdateExpression,
+	StringLiteral,
+	isVariableDeclaration,
+	VariableDeclaration,
+	VariableDeclarator,
+	isFunctionDeclaration,
+	ExpressionStatement,
+	isSequenceExpression,
+	SequenceExpression,
+	isUpdateExpression
+} from '@babel/types';
+import { IllegalStateException } from '@cqse/commons';
+
+const COVERAGE_OBJ_FUNCTION_NAME_PREFIX = 'cov_';
+
+type CoverageIncrement = {
+	type: 'statement' | 'branch' | 'function';
+	coverageObjectId: string;
+};
+
+type StatementCoverageIncrement = CoverageIncrement & {
+	type: 'statement';
+	statementId: number;
+};
+
+type BranchCoverageIncrement = CoverageIncrement & {
+	type: 'branch';
+	branchId: number;
+	locationId: number;
+};
+
+type FunctionCoverageIncrement = CoverageIncrement & {
+	type: 'function';
+	functionId: number;
+};
+
+function isIstanbulCoverageFunctionDeclaration(node: Node | undefined): string | undefined {
+	if (!isFunctionDeclaration(node)) {
+		return undefined;
+	}
+
+	const functionName = (node.id as Identifier).name;
+	if (functionName?.startsWith(COVERAGE_OBJ_FUNCTION_NAME_PREFIX)) {
+		return functionName;
+	} else {
+		return undefined;
+	}
+}
+
+/**
+ * Adds constants with the file id to the header of the file to process
+ * which are then used in the coverage broadcast functions as arguments.
+ *
+ * For example, `_$fid0` is introduced as `const _$fid0 = "6822844a804c1e9986ac4bd4a45b85893bde8b33";`
+ * based on
+ * ```
+ * function cov_oqh6rsgrd() {
+ *   var path = '/home/user/test/casestudies/plain-ts/dist/main.js';
+ *   var hash = '6822844a804c1e9986ac4bd4a45b85893bde8b33';
+ *   ....
+ * }
+ * ```
+ *
+ * And then used in the code, for example,
+ *```
+ * function s(e, r) {
+ *   _$stmtCov(_$fid0, 6)
+ *   ...
+ *   _$stmtCov(_$fid0, 18)
+ *   ...
+ * }
+ * ```.
+ */
+const fileIdMappingHandler = (() => {
+	const fileIdMap: Map<string, string> = new Map<string, string>();
+	let fileIdSeq = 0;
+
+	return {
+		enterPath(path: NodePath): void {
+			if (!isVariableDeclaration(path.node)) {
+				return;
+			}
+
+			const grandParentPath = path.parentPath?.parentPath;
+			const coverageFunctionName = isIstanbulCoverageFunctionDeclaration(grandParentPath?.node);
+			if (grandParentPath && coverageFunctionName) {
+				const declaration = path.node as VariableDeclaration;
+				if (declaration.declarations.length === 1) {
+					const declarator = declaration.declarations[0] as VariableDeclarator;
+					if ((declarator.id as Identifier).name === 'hash') {
```

Comment:
A bit more defensive

```suggestion
					if ('name' in declarator.id && declarator.id.name === 'hash') {
```